src/com/github/Kraken3/AFKPGC/BotDetector.java

```diff
@@ -0,0 +1,176 @@
+package com.github.Kraken3.AFKPGC;
+
+import java.util.Date;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.UUID;
+
+import org.bukkit.BanList;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+
+public class BotDetector implements Runnable {
+	static float currentTPS = 20;
+	static float acceptableTPS;
+	static float criticalTPSChange;
+	static long frequency; // how often this runs in ticks
+	float lastRoundTPS;
+	UUID topSuspect = null;
+	String topSuspectName = ""; // needs to be stored additionally,
+								// because retrieving the name off
+								// players after kicking them is only
+								// possible
+								// over deprecated/buggy methods
+	Location topSuspectsLocation = null;
+	UUID lastRoundSuspect = null;
+	String lastRoundSuspectName = "";
+	Location lastRoundSuspectsLocation = null;
+	static LinkedList<UUID> suspectedBotters = new LinkedList<UUID>();
+	static LinkedList<String> bannedPlayers = new LinkedList<String>();
+	// this is needed as a separated list, so we know the difference between
+	// players who were banned by AFKPGC and players who were banned for other
+	// reasons
+
+	static BanList banList = AFKPGC.plugin.getServer().getBanList(
+			BanList.Type.NAME); // ban after names not ips
+
+	public void run() {
+		if (!AFKPGC.enabled) {
+			return;
+		}
+		currentTPS = TpsReader.getTPS();
+		Map<UUID, LastActivity> lastActivities = LastActivity.lastActivities;
+		if (currentTPS < acceptableTPS) {
+			int smallestMovedDistance = 1024;
+			Iterator<Map.Entry<UUID, LastActivity>> entries = lastActivities

```

Comment:
Save an object, use the Set directly to iterate. Pattern:

    Set<Map.Entry<UUID, LastActivity>> entries = lastActivities.entrySet();
    for (Map.Entry<UUID, LastActivity> entry : entries) {

   }